File: packages/transport/src/tcp.ts
/**
 * Network printing over the JetDirect / RAW port.
 *
 * Practically every Ethernet or Wi-Fi label printer listens on TCP 9100 and
 * prints whatever it is handed. There is no protocol beyond "open, write,
 * close" — no acknowledgement and no status.
 *
 * That silence is why the timeouts matter. A printer that is switched off, or
 * one already holding a job from another workstation, leaves the connection
 * attempt hanging rather than refusing it, so without an explicit deadline a
 * print call can block until the operating system gives up minutes later.
 */

import { Socket } from 'node:net'

import { toBytes, type CommandEncoding } from './encode.js'
import { TransportError } from './errors.js'

/** The port essentially every raw-capable label printer listens on. */
export const RAW_PORT = 9100

export interface TcpOptions {
  /** Milliseconds to wait for the connection to be established. */
  connectTimeout?: number
  /** Milliseconds of inactivity tolerated once connected. */
  writeTimeout?: number
  encoding?: CommandEncoding
}

const DEFAULT_CONNECT_TIMEOUT = 5_000
const DEFAULT_WRITE_TIMEOUT = 15_000

/**
 * Send a compiled command stream to a networked printer.
 *
 * Resolves once the bytes have been flushed and the socket closed cleanly. The
 * close is significant: it is the only way the printer learns that the job has
 * ended, so a socket left open leaves the label unprinted.
 */
export function printTcp(
  host: string,
  port: number = RAW_PORT,
  payload: string | Uint8Array,
  options: TcpOptions = {},
): Promise<void> {
  const data = toBytes(payload, options.encoding)

  if (data.length === 0) {
    return Promise.reject(new TransportError('nothing to print', { host, port }))
  }

  const connectTimeout = options.connectTimeout ?? DEFAULT_CONNECT_TIMEOUT
  const writeTimeout = options.writeTimeout ?? DEFAULT_WRITE_TIMEOUT

  return new Promise<void>((resolve, reject) => {
    const socket = new Socket()
    let settled = false

    const finish = (error?: Error) => {
      if (settled) return
      settled = true
      socket.removeAllListeners()
      socket.destroy()
      error ? reject(error) : resolve()
    }

    const fail = (message: string, cause?: unknown) =>
      finish(new TransportError(message, { host, port, cause }))

    socket.setTimeout(connectTimeout)

    socket.once('timeout', () => {
      // The same event covers both phases, so the message reflects whichever
      // deadline is currently armed.
      fail(
        socket.connecting
          ? `timed out after ${connectTimeout} ms connecting to ${host}:${port}`
          : `timed out after ${writeTimeout} ms writing to ${host}:${port}`,
      )
    })

    socket.once('error', (error) => fail(`could not reach ${host}:${port}`, error))

    socket.connect(port, host, () => {
      socket.setTimeout(writeTimeout)
      socket.end(data, () => finish())
    })
  })
}

File: apps/desktop/src-tauri/src/printing/tcp.rs
//! Network printing over the JetDirect / RAW port.
//!
//! Practically every Ethernet or Wi-Fi label printer listens on TCP 9100 and
//! prints whatever it is handed. There is no protocol beyond "open, write,
//! close" — no acknowledgement and no status, which is why the timeouts matter:
//! a printer that is switched off accepts the connection attempt and then
//! simply never completes it.

use std::io::Write;
use std::net::{TcpStream, ToSocketAddrs};
use std::time::Duration;

use super::PrintError;

const CONNECT_TIMEOUT: Duration = Duration::from_secs(5);
const WRITE_TIMEOUT: Duration = Duration::from_secs(15);

pub fn print_tcp(host: &str, port: u16, data: &[u8]) -> Result<(), PrintError> {
    let address = (host, port)
        .to_socket_addrs()
        .map_err(|source| PrintError::Network {
            host: host.to_string(),
            port,
            source,
        })?
        .next()
        .ok_or_else(|| PrintError::Network {
            host: host.to_string(),
            port,
            source: std::io::Error::new(
                std::io::ErrorKind::NotFound,
                "host name did not resolve to an address",
            ),
        })?;

    let mut stream =
        TcpStream::connect_timeout(&address, CONNECT_TIMEOUT).map_err(|source| {
            PrintError::Network {
                host: host.to_string(),
                port,
                source,
            }
        })?;

    stream
        .set_write_timeout(Some(WRITE_TIMEOUT))
        .map_err(|source| PrintError::Network {
            host: host.to_string(),
            port,
            source,
        })?;

    stream.write_all(data)?;
    stream.flush()?;

    // Dropping the stream closes it, which is how the printer knows the job has
    // ended. There is nothing to read back.
    Ok(())
}
